spring-integration-core/src/main/java/org/springframework/integration/endpoint/Pausable.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.endpoint;
+
+import org.springframework.context.Lifecycle;
+
+/**
+ * Endpoints implementing this interface can be paused/resumed. A paused endpoint might
+ * still maintain a connection to an external system, but will not send or receive
+ * messages.
+ *
+ * @author Gary Russell
+ * @since 5.0.3
+ *
+ */
+public interface Pausable extends Lifecycle {
+
+	/**
+	 * Pause the endpoint.
+	 */
+	void pause();
```

Comment:
The contract of the Kafka `Consumer` is like:
```
public void pause(Collection<TopicPartition> partitions);

public void resume(Collection<TopicPartition> partitions);
```

Maybe we can make this as `Supplier<T>` as well?
Or at least as an overloaded variants:
```
default void pause(Supplier<T> pauseOn) {
    pause();
}
```
?

Reply:
I don't think we need to support partial pause; this is what I have...

```java
	@Override
	public void onApplicationEvent(ListenerContainerIdleEvent event) {
		Consumer<?, ?> consumer = event.getConsumer();
		pauseIfNecessary(consumer);
		if (this.resumePending.getAndSet(false) && this.paused.getAndSet(false)) {
			consumer.resume(consumer.paused());
		}
	}

	private void pauseIfNecessary(Consumer<?, ?> consumer) {
		if (this.pausePending.getAndSet(false) && !this.paused.getAndSet(true)) {
			consumer.pause(consumer.assignment());
		}
	}
```

Reply:
OK. Merging...